File: packages/typescript/runtime/treadle.ts
import { create, toBinary } from "@bufbuild/protobuf";
import { blake3 } from "@noble/hashes/blake3.js";
import {
  TreadleArgvSchema,
  TreadleCheckClass,
  TreadleCheckSchema,
  TreadleDefinitionSchema,
  TreadleDeterminismClass,
  TreadleEnvEntrySchema,
  TreadleIsolationHintsSchema,
  TreadleJobSchema,
  TreadleMatrixValueSchema,
  TreadleNetworkAccess,
  TreadlePlatformSchema,
  TreadleRetrySchema,
  TreadleSecretRefSchema,
  TreadleSecretTier,
  TreadleServiceContainerSchema,
  TreadleTargetEnvironmentSchema,
  TreadleTriggerKind,
  TreadleTriggerSchema,
  type TreadleCheck,
  type TreadleDefinition,
  type TreadleEnvEntry,
  type TreadleServiceContainer,
  type TreadleTrigger,
} from "./treadle_pb.js";

export const TREADLE_DEFINITION_FORMAT_VERSION = 1;

export class TreadleDefinitionError extends Error {
  override name = "TreadleDefinitionError";
}

/** Validate, normalize, and protobuf-encode a v1 treadle definition canonically. */
export function canonicalTreadleDefinitionBytes(
  definition: TreadleDefinition,
): Uint8Array {
  return toBinary(TreadleDefinitionSchema, canonicalDefinition(definition), {
    writeUnknownFields: false,
  });
}

/** BLAKE3 content address of the canonical protobuf bytes. */
export function treadleDefinitionBlake3(definition: TreadleDefinition): Uint8Array {
  return blake3(canonicalTreadleDefinitionBytes(definition));
}

function canonicalDefinition(definition: TreadleDefinition): TreadleDefinition {
  if (definition.formatVersion !== TREADLE_DEFINITION_FORMAT_VERSION) {
    throw new TreadleDefinitionError(
      `unsupported treadle definition format version ${definition.formatVersion}; ` +
        `migrate to version ${TREADLE_DEFINITION_FORMAT_VERSION} before reading`,
    );
  }
  nonEmpty("pipeline name", definition.name);

  const secretRefs = definition.secretRefs
    .map((secret) => {
      identifier("secret", secret.name);
      noNul("secret provider", secret.provider);
      if (![TreadleSecretTier.STANDARD, TreadleSecretTier.TRUSTED_RUNNER_ONLY].includes(secret.tier)) {
        throw new TreadleDefinitionError(
          `secret ${JSON.stringify(secret.name)} has an invalid tier`,
        );
      }
      return create(TreadleSecretRefSchema, secret);
    })
    .sort((left, right) => byteCompare(left.name, right.name));
  uniqueNames("secret declaration", secretRefs);
  const secretNames = new Set(secretRefs.map((secret) => secret.name));

  const services = definition.services
    .map((service) => canonicalService(service, secretNames))
    .sort((left, right) => byteCompare(left.name, right.name));
  uniqueNames("service", services);
  const serviceNames = new Set(services.map((service) => service.name));

  const jobs = definition.jobs
    .map((job) => {
      identifier("job", job.name);
      const matrix = job.matrix
        .map((binding) => {
          identifier("matrix dimension", binding.name);
          noNul("matrix value", binding.value);
          if (binding.value.includes("${") || binding.value.includes("{{")) {
            throw new TreadleDefinitionError(
              `job ${JSON.stringify(job.name)} matrix value ${JSON.stringify(binding.value)} looks unresolved`,
            );
          }
          return create(TreadleMatrixValueSchema, binding);
        })
        .sort((left, right) => byteCompare(left.name, right.name));
      uniqueNames("matrix dimension", matrix);

      const checks = job.checks
        .map((check) => canonicalCheck(check, serviceNames, secretNames))
        .sort((left, right) => byteCompare(left.name, right.name));
      uniqueNames("check", checks);
      if (checks.length === 0) {
        throw new TreadleDefinitionError(`job ${JSON.stringify(job.name)} has no checks`);
      }
      return create(TreadleJobSchema, { name: job.name, matrix, checks });
    })
    .sort((left, right) => byteCompare(left.name, right.name));
  if (jobs.length === 0) {
    throw new TreadleDefinitionError("pipeline has no jobs");
  }
  uniqueNames("job", jobs);

  return create(TreadleDefinitionSchema, {
    formatVersion: TREADLE_DEFINITION_FORMAT_VERSION,
    name: definition.name,
    jobs,
    services,
    secretRefs,
  });
}

function canonicalCheck(
  check: TreadleCheck,
  serviceNames: Set<string>,
  secretNames: Set<string>,
): TreadleCheck {
  identifier("check", check.name);
  nonEmpty("check command", check.command);
  check.args.forEach((arg) => noNul("check argument", arg));
  if (![TreadleCheckClass.REQUIRED, TreadleCheckClass.ADVISORY, TreadleCheckClass.INFORMATIONAL].includes(check.class)) {
    throw new TreadleDefinitionError(`check ${JSON.stringify(check.name)} has an invalid class`);
  }
  if (![
    TreadleDeterminismClass.DETERMINISTIC,
    TreadleDeterminismClass.NONDETERMINISTIC,
  ].includes(check.determinismClass)) {
    throw new TreadleDefinitionError(
      `check ${JSON.stringify(check.name)} has an invalid determinismClass`,
    );
  }
  positiveUint32("timeoutSeconds", check.timeoutSeconds);
  relativePath("workingDirectory", check.workingDirectory, true);

  if (check.targetEnvironment === undefined) {
    throw new TreadleDefinitionError(
      `check ${JSON.stringify(check.name)} omits targetEnvironment`,
    );
  }
  ociImageDigest("target environment OCI image digest", check.targetEnvironment.ociImageDigest);
  if (check.targetEnvironment.platform === undefined) {
    throw new TreadleDefinitionError(
      `check ${JSON.stringify(check.name)} omits targetEnvironment.platform`,
    );
  }
  platformValue("target platform os", check.targetEnvironment.platform.os);
  platformValue("target platform arch", check.targetEnvironment.platform.arch);
  const targetEnvironment = create(TreadleTargetEnvironmentSchema, {
    ociImageDigest: check.targetEnvironment.ociImageDigest,
    platform: create(TreadlePlatformSchema, check.targetEnvironment.platform),
  });

  const env = canonicalEnv(check.env, secretNames);
  const serviceDependencies = sortedUnique("service dependency", check.serviceDependencies);
  for (const dependency of serviceDependencies) {
    if (!serviceNames.has(dependency)) {
      throw new TreadleDefinitionError(
        `check ${JSON.stringify(check.name)} references undeclared service ${JSON.stringify(dependency)}`,
      );
    }
  }

  if (check.retry === undefined) {
    throw new TreadleDefinitionError(`check ${JSON.stringify(check.name)} omits retry`);
  }
  const flakeSignatures = sortedUnique("flake signature", check.retry.flakeSignatures);
  flakeSignatures.forEach((signature) => nonEmpty("flake signature", signature));
  const retry = create(TreadleRetrySchema, {
    maxRetries: uint32("maxRetries", check.retry.maxRetries),
    flakeSignatures,
  });

  const cachePaths = sortedUnique("cache path", check.cachePaths);
  cachePaths.forEach((path) => relativePath("cache path", path, false));

  if (check.isolation === undefined) {
    throw new TreadleDefinitionError(`check ${JSON.stringify(check.name)} omits isolation`);
  }
  noNul("isolation profile", check.isolation.profile);
  const cpuMillis = uint32("cpuMillis", check.isolation.cpuMillis);
  const processLimit = uint32("processLimit", check.isolation.processLimit);
  if (
    typeof check.isolation.memoryBytes !== "bigint" ||
    check.isolation.memoryBytes < 0n ||
    check.isolation.memoryBytes > 0xffff_ffff_ffff_ffffn
  ) {
    throw new TreadleDefinitionError("memoryBytes must be a uint64");
  }
  if (![
    TreadleNetworkAccess.UNSPECIFIED,
    TreadleNetworkAccess.NONE,
    TreadleNetworkAccess.SERVICES_ONLY,
    TreadleNetworkAccess.FULL,
  ].includes(check.isolation.networkAccess)) {
    throw new TreadleDefinitionError(
      `check ${JSON.stringify(check.name)} has invalid networkAccess`,
    );
  }
  const isolation = create(TreadleIsolationHintsSchema, {
    ...check.isolation,
    cpuMillis,
    processLimit,
  });

  const triggers = check.triggers
    .map((trigger) => canonicalTrigger(check.name, trigger))
    .sort((left, right) => left.kind - right.kind || byteCompare(left.cronExpression, right.cronExpression));
  if (triggers.length === 0) {
    throw new TreadleDefinitionError(`check ${JSON.stringify(check.name)} has no triggers`);
  }
  const triggerKeys = triggers.map((trigger) => `${trigger.kind}\u0000${trigger.cronExpression}`);
  if (new Set(triggerKeys).size !== triggerKeys.length) {
    throw new TreadleDefinitionError(`check ${JSON.stringify(check.name)} has a duplicate trigger`);
  }

  return create(TreadleCheckSchema, {
    name: check.name,
    command: check.command,
    args: [...check.args],
    class: check.class,
    timeoutSeconds: check.timeoutSeconds,
    env,
    workingDirectory: check.workingDirectory,
    serviceDependencies,
    retry,
    cachePaths,
    isolation,
    triggers,
    supersedeOlderRuns: check.supersedeOlderRuns,
    targetEnvironment,
    determinismClass: check.determinismClass,
  });
}

function canonicalService(
  service: TreadleServiceContainer,
  secretNames: Set<string>,
): TreadleServiceContainer {
  identifier("service", service.name);
  nonEmpty("service image", service.image);
  ociImageDigest("service OCI image digest", service.ociImageDigest);
  const ports = [...service.ports].sort((left, right) => left - right);
  ports.forEach((port) => {
    if (!Number.isInteger(port) || port < 1 || port > 65535) {
      throw new TreadleDefinitionError(`service ${JSON.stringify(service.name)} has invalid port ${port}`);
    }
  });
  if (new Set(ports).size !== ports.length) {
    throw new TreadleDefinitionError(`service ${JSON.stringify(service.name)} has a duplicate port`);
  }
  let readiness;
  if (service.readiness !== undefined) {
    nonEmpty("service readiness command", service.readiness.command);
    service.readiness.args.forEach((arg) => noNul("service readiness argument", arg));
    readiness = create(TreadleArgvSchema, {
      command: service.readiness.command,
      args: [...service.readiness.args],
    });
  }
  return create(TreadleServiceContainerSchema, {
    name: service.name,
    image: service.image,
    ports,
    env: canonicalEnv(service.env, secretNames),
    readiness,
    ociImageDigest: service.ociImageDigest,
  });
}

function canonicalEnv(env: TreadleEnvEntry[], secretNames: Set<string>): TreadleEnvEntry[] {
  const result = env
    .map((entry) => {
      nonEmpty("environment name", entry.name);
      if (entry.source.case === "literalValue") {
        noNul("environment literal", entry.source.value);
      } else if (entry.source.case === "secretRef") {
        if (!secretNames.has(entry.source.value)) {
          throw new TreadleDefinitionError(
            `environment references undeclared secret ${JSON.stringify(entry.source.value)}`,
          );
        }
      } else {
        throw new TreadleDefinitionError(
          `environment ${JSON.stringify(entry.name)} has no source`,
        );
      }
      return create(TreadleEnvEntrySchema, entry);
    })
    .sort((left, right) => byteCompare(left.name, right.name));
  uniqueNames("environment entry", result);
  return result;
}

function canonicalTrigger(checkName: string, trigger: TreadleTrigger): TreadleTrigger {
  if (
    (trigger.kind === TreadleTriggerKind.PUSH || trigger.kind === TreadleTriggerKind.MANUAL) &&
    trigger.cronExpression === ""
  ) {
    return create(TreadleTriggerSchema, trigger);
  }
  if (trigger.kind === TreadleTriggerKind.CRON && validCron(trigger.cronExpression)) {
    return create(TreadleTriggerSchema, trigger);
  }
  throw new TreadleDefinitionError(`check ${JSON.stringify(checkName)} has an invalid trigger`);
}

function uniqueNames(kind: string, values: Array<{ name: string }>): void {
  if (new Set(values.map((value) => value.name)).size !== values.length) {
    throw new TreadleDefinitionError(`duplicate ${kind} name`);
  }
}

function sortedUnique(kind: string, values: string[]): string[] {
  const result = [...values].sort(byteCompare);
  if (new Set(result).size !== result.length) {
    throw new TreadleDefinitionError(`duplicate ${kind}`);
  }
  return result;
}

function identifier(kind: string, value: string): void {
  if (
    value.length === 0 ||
    ![...value].every((character) =>
      (character >= "a" && character <= "z") ||
      (character >= "0" && character <= "9") ||
      "._:-".includes(character),
    )
  ) {
    throw new TreadleDefinitionError(
      `${kind} name ${JSON.stringify(value)} must match [a-z0-9._:-]+`,
    );
  }
}

function nonEmpty(kind: string, value: string): void {
  if (value.length === 0) throw new TreadleDefinitionError(`${kind} must not be empty`);
  noNul(kind, value);
}

function noNul(kind: string, value: string): void {
  if (value.includes("\u0000")) {
    throw new TreadleDefinitionError(`${kind} must not contain NUL`);
  }
}

function ociImageDigest(kind: string, value: string): void {
  if (!/^sha256:[0-9a-f]{64}$/.test(value)) {
    throw new TreadleDefinitionError(
      `${kind} must be sha256: followed by 64 lowercase hexadecimal digits`,
    );
  }
}

function platformValue(kind: string, value: string): void {
  if (!/^[a-z0-9]+$/.test(value)) {
    throw new TreadleDefinitionError(`${kind} must match [a-z0-9]+`);
  }
}

function positiveUint32(kind: string, value: number): void {
  if (uint32(kind, value) === 0) {
    throw new TreadleDefinitionError(`${kind} must be a positive uint32`);
  }
}

function uint32(kind: string, value: number): number {
  if (!Number.isInteger(value) || value < 0 || value > 0xffff_ffff) {
    throw new TreadleDefinitionError(`${kind} must be a uint32`);
  }
  return value;
}

function relativePath(kind: string, value: string, allowEmpty: boolean): void {
  if (value.length === 0 && allowEmpty) return;
  const segments = value.split("/");
  if (
    value.length === 0 ||
    value.startsWith("/") ||
    value.includes("\\") ||
    /^[A-Za-z]:/.test(value) ||
    segments.some((segment) => segment === "" || segment === "." || segment === "..")
  ) {
    throw new TreadleDefinitionError(
      `${kind} ${JSON.stringify(value)} must be a normalized repository-relative path`,
    );
  }
  noNul(kind, value);
}

function validCron(value: string): boolean {
  const fields = value.split(" ");
  return fields.length === 5 && fields.every((field) =>
    field.length > 0 && [...field].every((character) =>
      (character >= "0" && character <= "9") || "*,-/".includes(character),
    ),
  );
}

const utf8 = new TextEncoder();

function byteCompare(left: string, right: string): number {
  const leftBytes = utf8.encode(left);
  const rightBytes = utf8.encode(right);
  const length = Math.min(leftBytes.length, rightBytes.length);
  for (let index = 0; index < length; index += 1) {
    const difference = leftBytes[index]! - rightBytes[index]!;
    if (difference !== 0) return difference;
  }
  return leftBytes.length - rightBytes.length;
}

File: src/treadle.rs
// SPDX-License-Identifier: Apache-2.0
//! Canonical encoding for the signed treadle definition contract.

use std::collections::BTreeSet;

use prost::Message;
use thiserror::Error;

use crate::heddle::api::v1alpha1::{
    TreadleCheck, TreadleCheckClass, TreadleDefinition, TreadleEnvEntry, TreadleJob,
    TreadleNetworkAccess, TreadleServiceContainer, TreadleTrigger, TreadleTriggerKind,
    treadle_env_entry,
};

/// The only treadle definition format accepted by this release.
pub const TREADLE_DEFINITION_FORMAT_VERSION: u32 = 1;

/// Canonical treadle definition validation/decoding errors.
#[derive(Debug, Error)]
pub enum TreadleDefinitionError {
    #[error(
        "unsupported treadle definition format version {actual}; migrate to version {expected} before reading"
    )]
    UnsupportedVersion { actual: u32, expected: u32 },
    #[error("invalid treadle definition: {0}")]
    Invalid(String),
    #[error("invalid treadle protobuf: {0}")]
    Decode(#[from] prost::DecodeError),
    #[error("treadle definition bytes are not canonical")]
    NonCanonicalBytes,
}

/// Validate, normalize, and encode a definition using the v1 canonical wire rule.
pub fn canonical_treadle_definition_bytes(
    definition: &TreadleDefinition,
) -> Result<Vec<u8>, TreadleDefinitionError> {
    let normalized = canonical_treadle_definition(definition)?;
    Ok(normalized.encode_to_vec())
}

/// Return the BLAKE3 content address of the v1 canonical protobuf bytes.
pub fn treadle_definition_blake3(
    definition: &TreadleDefinition,
) -> Result<[u8; 32], TreadleDefinitionError> {
    Ok(*blake3::hash(&canonical_treadle_definition_bytes(definition)?).as_bytes())
}

/// Decode only a byte-exact canonical current-version definition.
///
/// This rejects alternate protobuf encodings, unknown fields, unsorted repeated
/// sets, invalid definitions, and old/future versions. Migration is a separate,
/// explicit operation; this reader never provides a dual-read fallback.
pub fn decode_canonical_treadle_definition(
    bytes: &[u8],
) -> Result<TreadleDefinition, TreadleDefinitionError> {
    let decoded = TreadleDefinition::decode(bytes)?;
    let normalized = canonical_treadle_definition(&decoded)?;
    if normalized.encode_to_vec() != bytes {
        return Err(TreadleDefinitionError::NonCanonicalBytes);
    }
    Ok(normalized)
}

fn canonical_treadle_definition(
    definition: &TreadleDefinition,
) -> Result<TreadleDefinition, TreadleDefinitionError> {
    if definition.format_version != TREADLE_DEFINITION_FORMAT_VERSION {
        return Err(TreadleDefinitionError::UnsupportedVersion {
            actual: definition.format_version,
            expected: TREADLE_DEFINITION_FORMAT_VERSION,
        });
    }
    non_empty("pipeline name", &definition.name)?;

    let mut normalized = definition.clone();
    normalized
        .secret_refs
        .sort_by(|left, right| left.name.cmp(&right.name));
    normalized
        .services
        .sort_by(|left, right| left.name.cmp(&right.name));
    normalized
        .jobs
        .sort_by(|left, right| left.name.cmp(&right.name));
    if normalized.jobs.is_empty() {
        return invalid("pipeline has no jobs".into());
    }

    let secret_names = unique_named(
        "secret declaration",
        normalized
            .secret_refs
            .iter()
            .map(|secret| secret.name.as_str()),
    )?;
    for secret in &normalized.secret_refs {
        identifier("secret", &secret.name)?;
        no_nul("secret provider", &secret.provider)?;
    }

    let service_names = unique_named(
        "service",
        normalized
            .services
            .iter()
            .map(|service| service.name.as_str()),
    )?;
    for service in &mut normalized.services {
        normalize_service(service, &secret_names)?;
    }

    unique_named("job", normalized.jobs.iter().map(|job| job.name.as_str()))?;
    for job in &mut normalized.jobs {
        normalize_job(job, &service_names, &secret_names)?;
    }

    Ok(normalized)
}

fn normalize_job(
    job: &mut TreadleJob,
    service_names: &BTreeSet<String>,
    secret_names: &BTreeSet<String>,
) -> Result<(), TreadleDefinitionError> {
    identifier("job", &job.name)?;
    job.matrix.sort_by(|left, right| left.name.cmp(&right.name));
    unique_named(
        "matrix dimension",
        job.matrix.iter().map(|value| value.name.as_str()),
    )?;
    for value in &job.matrix {
        identifier("matrix dimension", &value.name)?;
        no_nul("matrix value", &value.value)?;
        if value.value.contains("${") || value.value.contains("{{") {
            return invalid(format!(
                "job {:?} matrix value {:?} looks unresolved",
                job.name, value.value
            ));
        }
    }

    job.checks.sort_by(|left, right| left.name.cmp(&right.name));
    unique_named("check", job.checks.iter().map(|check| check.name.as_str()))?;
    if job.checks.is_empty() {
        return invalid(format!("job {:?} has no checks", job.name));
    }
    for check in &mut job.checks {
        normalize_check(check, service_names, secret_names)?;
    }
    Ok(())
}

fn normalize_check(
    check: &mut TreadleCheck,
    service_names: &BTreeSet<String>,
    secret_names: &BTreeSet<String>,
) -> Result<(), TreadleDefinitionError> {
    identifier("check", &check.name)?;
    non_empty("check command", &check.command)?;
    for arg in &check.args {
        no_nul("check argument", arg)?;
    }
    match TreadleCheckClass::try_from(check.class) {
        Ok(
            TreadleCheckClass::Required
            | TreadleCheckClass::Advisory
            | TreadleCheckClass::Informational,
        ) => {}
        _ => return invalid(format!("check {:?} has an invalid class", check.name)),
    }
    if check.timeout_seconds == 0 {
        return invalid(format!(
            "check {:?} timeout_seconds must be positive",
            check.name
        ));
    }
    normalize_env(&mut check.env, secret_names)?;
    relative_path("working_directory", &check.working_directory, true)?;

    check.service_dependencies.sort();
    unique_strings("service dependency", &check.service_dependencies)?;
    for dependency in &check.service_dependencies {
        if !service_names.contains(dependency) {
            return invalid(format!(
                "check {:?} references undeclared service {:?}",
                check.name, dependency
            ));
        }
    }

    let retry = check.retry.as_mut().ok_or_else(|| {
        TreadleDefinitionError::Invalid(format!("check {:?} omits retry", check.name))
    })?;
    retry.flake_signatures.sort();
    unique_strings("flake signature", &retry.flake_signatures)?;
    for signature in &retry.flake_signatures {
        non_empty("flake signature", signature)?;
    }

    check.cache_paths.sort();
    unique_strings("cache path", &check.cache_paths)?;
    for path in &check.cache_paths {
        relative_path("cache path", path, false)?;
    }

    let isolation = check.isolation.as_ref().ok_or_else(|| {
        TreadleDefinitionError::Invalid(format!("check {:?} omits isolation", check.name))
    })?;
    no_nul("isolation profile", &isolation.profile)?;
    if TreadleNetworkAccess::try_from(isolation.network_access).is_err() {
        return invalid(format!("check {:?} has invalid network_access", check.name));
    }

    check.triggers.sort_by(|left, right| {
        (left.kind, left.cron_expression.as_str())
            .cmp(&(right.kind, right.cron_expression.as_str()))
    });
    if check.triggers.is_empty() {
        return invalid(format!("check {:?} has no triggers", check.name));
    }
    let mut trigger_keys = BTreeSet::new();
    for trigger in &check.triggers {
        validate_trigger(check, trigger)?;
        if !trigger_keys.insert((trigger.kind, trigger.cron_expression.clone())) {
            return invalid(format!("check {:?} has a duplicate trigger", check.name));
        }
    }
    Ok(())
}

fn validate_trigger(
    check: &TreadleCheck,
    trigger: &TreadleTrigger,
) -> Result<(), TreadleDefinitionError> {
    match TreadleTriggerKind::try_from(trigger.kind) {
        Ok(TreadleTriggerKind::Push | TreadleTriggerKind::Manual)
            if trigger.cron_expression.is_empty() =>
        {
            Ok(())
        }
        Ok(TreadleTriggerKind::Cron) if valid_cron(&trigger.cron_expression) => Ok(()),
        _ => invalid(format!("check {:?} has an invalid trigger", check.name)),
    }
}

fn normalize_service(
    service: &mut TreadleServiceContainer,
    secret_names: &BTreeSet<String>,
) -> Result<(), TreadleDefinitionError> {
    identifier("service", &service.name)?;
    non_empty("service image", &service.image)?;
    service.ports.sort_unstable();
    if service
        .ports
        .iter()
        .any(|port| *port == 0 || *port > u16::MAX.into())
    {
        return invalid(format!("service {:?} has an invalid port", service.name));
    }
    if service.ports.windows(2).any(|pair| pair[0] == pair[1]) {
        return invalid(format!("service {:?} has a duplicate port", service.name));
    }
    normalize_env(&mut service.env, secret_names)?;
    if let Some(readiness) = &service.readiness {
        non_empty("service readiness command", &readiness.command)?;
        for arg in &readiness.args {
            no_nul("service readiness argument", arg)?;
        }
    }
    Ok(())
}

fn normalize_env(
    env: &mut [TreadleEnvEntry],
    secret_names: &BTreeSet<String>,
) -> Result<(), TreadleDefinitionError> {
    env.sort_by(|left, right| left.name.cmp(&right.name));
    unique_named(
        "environment entry",
        env.iter().map(|entry| entry.name.as_str()),
    )?;
    for entry in env {
        non_empty("environment name", &entry.name)?;
        match entry.source.as_ref() {
            Some(treadle_env_entry::Source::LiteralValue(value)) => {
                no_nul("environment literal", value)?;
            }
            Some(treadle_env_entry::Source::SecretRef(name)) if secret_names.contains(name) => {}
            Some(treadle_env_entry::Source::SecretRef(name)) => {
                return invalid(format!("environment references undeclared secret {name:?}"));
            }
            None => return invalid(format!("environment {:?} has no source", entry.name)),
        }
    }
    Ok(())
}

fn unique_named<'a>(
    kind: &str,
    names: impl IntoIterator<Item = &'a str>,
) -> Result<BTreeSet<String>, TreadleDefinitionError> {
    let mut seen = BTreeSet::new();
    for name in names {
        if !seen.insert(name.to_owned()) {
            return invalid(format!("duplicate {kind} name {name:?}"));
        }
    }
    Ok(seen)
}

fn unique_strings(kind: &str, values: &[String]) -> Result<(), TreadleDefinitionError> {
    if values.windows(2).any(|pair| pair[0] == pair[1]) {
        return invalid(format!("duplicate {kind}"));
    }
    Ok(())
}

fn identifier(kind: &str, value: &str) -> Result<(), TreadleDefinitionError> {
    if value.is_empty()
        || !value.bytes().all(|byte| {
            byte.is_ascii_lowercase() || byte.is_ascii_digit() || b"._:-".contains(&byte)
        })
    {
        return invalid(format!("{kind} name {value:?} must match [a-z0-9._:-]+"));
    }
    Ok(())
}

fn non_empty(kind: &str, value: &str) -> Result<(), TreadleDefinitionError> {
    if value.is_empty() {
        return invalid(format!("{kind} must not be empty"));
    }
    no_nul(kind, value)
}

fn no_nul(kind: &str, value: &str) -> Result<(), TreadleDefinitionError> {
    if value.contains('\0') {
        return invalid(format!("{kind} must not contain NUL"));
    }
    Ok(())
}

fn relative_path(kind: &str, value: &str, allow_empty: bool) -> Result<(), TreadleDefinitionError> {
    if value.is_empty() {
        return if allow_empty {
            Ok(())
        } else {
            invalid(format!("{kind} must not be empty"))
        };
    }
    if value.starts_with('/')
        || value.contains('\\')
        || value.as_bytes().get(1) == Some(&b':')
        || value
            .split('/')
            .any(|segment| segment.is_empty() || segment == "." || segment == "..")
    {
        return invalid(format!(
            "{kind} {value:?} must be a normalized repository-relative path"
        ));
    }
    no_nul(kind, value)
}

fn valid_cron(value: &str) -> bool {
    let fields = value.split(' ').collect::<Vec<_>>();
    fields.len() == 5
        && fields.iter().all(|field| {
            !field.is_empty()
                && field
                    .bytes()
                    .all(|byte| byte.is_ascii_digit() || b"*,-/".contains(&byte))
        })
}

fn invalid<T>(message: String) -> Result<T, TreadleDefinitionError> {
    Err(TreadleDefinitionError::Invalid(message))
}
